### PR TITLE
Lab8

### PR DESCRIPTION
**Pull request by Kozak24**

### CODE REVIEW

File: app/src/main/java/kozak/labs/ApplicationEx.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        mFragmentNavigation = fragmentNavigation;
+    }
+
+    public static Context getContext() {
+        return mContext;
+    }
+
+    public static void setContext(Context Context) {
+        mContext = Context;
+    }
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/ApplicationEx.java]: You should never access ApplicationEx as a static object. Same goes for all methods above

File: app/src/main/java/kozak/labs/FragmentNavigation/FragmentNavigation.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    public FragmentNavigation(FragmentManager fragmentManager) {
+        this.fragmentManager = fragmentManager;
+    }
+
+    public void setFragment(final Fragment fragment) {
+        fragmentManager.beginTransaction()
+                .replace( R.id.fragment_container, fragment)
+                .addToBackStack(null)
+                .commit();
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/FragmentNavigation/FragmentNavigation.java]: this method should also accept `final boolean addToBackStack`

Comment (MarkoKuspis) [reply app/src/main/java/kozak/labs/FragmentNavigation/FragmentNavigation.java]: final FragmentTransaction transaction = fragmentManager.beginTransaction()
                .replace( R.id.fragment_container, fragment);

if (addToBackStack) {
    transaction.addToBackStack(null)
}
transaction.commit();

File: app/src/main/java/kozak/labs/Fragments/FavoritesFragment.java [hunk 3]
```diff
@@ -80,15 +67,16 @@ public void onCharacterClick(Character character) {
                     ListItemFragment listItemFragment = new ListItemFragment();
                     listItemFragment.setArguments(bundle);
 
-                    mainActivity.setFragment(listItemFragment);
+                    ApplicationEx.getFragmentNavigation().setFragment(listItemFragment);
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/Fragments/FavoritesFragment.java]: `((ApplicationEx) getActivity().getApplication()).getFragmentNavigation().setFragment(...)`

File: app/src/main/java/kozak/labs/Fragments/ListFragment.java [hunk 4]
```diff
@@ -103,23 +82,29 @@ public void onCharacterClick(Character character) {
...
+    public void noData() {
         adapter.setItems(null);
         adapter.notifyDataSetChanged();
         noDataTextView.setVisibility(View.VISIBLE);
     }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        mPresenter.onDetachView();
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/Fragments/ListFragment.java]: 👍

File: app/src/main/java/kozak/labs/Presenter/ListItemPresenter.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+public class ListItemPresenter implements ListItemFragmentContract.Presenter {
+    private ListItemFragmentContract.View mView;
+    private ListItemFragmentContract.Model mModel;
+
+    public ListItemPresenter(ListItemFragmentContract.View mView) {
+        this.mView = mView;
+        this.mModel = new ListItemModel();
+    }
+
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/Presenter/ListItemPresenter.java]:
```
public onResume() {
   final Character character = ApplicationEx.getCharacter();
   // TODO: go to shared pref and check if character is favourite
   ...
   view.showCharacter(character, true);
}
```

File: app/src/main/java/kozak/labs/MVPInterfaces/FavoritesFragmentContract.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package kozak.labs.MVPInterfaces;
+
+import java.util.List;
+
+import kozak.labs.Entity.Character;
+
+public interface FavoritesFragmentContract {
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/MVPInterfaces/FavoritesFragmentContract.java]: This should be renamed to `FavoriteCharactersContract`

File: app/src/main/java/kozak/labs/MVPInterfaces/FavoritesFragmentContract.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    interface View {
+        void displayItems(List<Character> characterList);
+    }
+
+    interface Presenter {
+        void loadData();
+    }
+
+    interface Model {
+        List<Character> getPreferences();
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/MVPInterfaces/FavoritesFragmentContract.java]:
```suggestion
        List<Character> getFavoriteCharacters();
```

File: app/src/main/java/kozak/labs/MVPInterfaces/ListFragmentContract.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package kozak.labs.MVPInterfaces;
+
+import java.util.List;
+
+import kozak.labs.Entity.Character;
+
+public interface ListFragmentContract {
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/MVPInterfaces/ListFragmentContract.java]:
```suggestion
public interface AnimeCharactersContract {
```

File: app/src/main/java/kozak/labs/MVPInterfaces/ListFragmentContract.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    interface View {
+        void noData();
+        void displayItems(List<Character> characterList);
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/MVPInterfaces/ListFragmentContract.java]: void displayCharacters(List<Character> characters);

File: app/src/main/java/kozak/labs/MVPInterfaces/ListFragmentContract.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    interface Presenter {
+        void onDetachView();
+        void loadData();
+    }
+
+    interface Model {
+        List<Character> makeCall();
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/MVPInterfaces/ListFragmentContract.java]:
```suggestion
        List<Character> getCharacters();
```

File: app/src/main/java/kozak/labs/MVPInterfaces/ListItemFragmentContract.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package kozak.labs.MVPInterfaces;
+
+public interface ListItemFragmentContract {
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/MVPInterfaces/ListItemFragmentContract.java]:
```suggestion
public interface CharacterDetailsContract {
```

File: app/src/main/java/kozak/labs/MVPInterfaces/ListItemFragmentContract.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    interface View {
+        void setFavoriteImage(boolean isFavorite);
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/MVPInterfaces/ListItemFragmentContract.java]:
```suggestion
        void showCharacter(final Character character, final boolean isFavorite);
```

File: app/src/main/java/kozak/labs/MVPInterfaces/ListItemFragmentContract.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+    }
+
+    interface Presenter {
+        void makeFavorite();
+        void isFavorite();
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/MVPInterfaces/ListItemFragmentContract.java]:
```suggestion
void changeFavoriteState();
```

File: app/src/main/java/kozak/labs/ApplicationEx.java [hunk 14]
```diff
@@ -37,10 +35,11 @@ public static void setCharacter(Character character) {
...
     }
 
-
-
     @Override
     public void onCreate() {
         super.onCreate();
+        mContext = getApplicationContext();
+        /*mFragmentNavigation = new FragmentNavigation(
+                ((MainActivity) mContext).getSupportFragmentManager());*/
```
Comment (Kozak24) [inline app/src/main/java/kozak/labs/ApplicationEx.java]: This comment deleted in newer commit

File: app/src/main/java/kozak/labs/ApplicationEx.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        return mFragmentNavigation;
+    }
+
+    public void setFragmentNavigation(FragmentNavigation fragmentNavigation) {
+        mFragmentNavigation = fragmentNavigation;
+    }
+
+    public Context getContext() {
+        return mContext;
+    }
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/ApplicationEx.java]: Drop this one

File: app/src/main/java/kozak/labs/ApplicationEx.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+import kozak.labs.Entity.Character;
+import kozak.labs.FragmentNavigation.FragmentNavigation;
+import kozak.labs.Model.DetailsModel;
+import kozak.labs.Model.FavoritesModel;
+import kozak.labs.Model.ListModel;
+
+public class ApplicationEx extends Application {
+    private Context mContext;
+    private static Character mCharacter;
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/ApplicationEx.java]: currentCharacter

File: app/src/main/java/kozak/labs/Fragments/DetailsFragment.java [hunk 17]
```diff
@@ -52,23 +53,28 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
         ButterKnife.bind(this, view);
         if(getArguments() != null) {
             character = (Character) getArguments().getSerializable(Constants.ARG_TITLE);
-            displayCharacter();
+            ApplicationEx.setCharacter(character);
         }
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/Fragments/DetailsFragment.java]: Drop block above

File: app/src/main/java/kozak/labs/Fragments/FavoritesFragment.java [hunk 18]
```diff
@@ -77,19 +63,28 @@ public void onCharacterClick(Character character) {
                     Bundle bundle = new Bundle();
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/Fragments/FavoritesFragment.java]:
```
presenter.characterSelecter(character)
```

File: app/src/main/java/kozak/labs/FragmentNavigation/FragmentNavigation.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+    public void setFragment(final Fragment fragment, final boolean addToBackStack) {
+        final FragmentTransaction transaction = fragmentManager.beginTransaction()
+                .replace( R.id.fragment_container, fragment);
+
+        if (addToBackStack) {
+            transaction.addToBackStack( null );
+        }
+        transaction.commit();
+    }
```
Comment (MarkoKuspis) [inline app/src/main/java/kozak/labs/FragmentNavigation/FragmentNavigation.java]:
```
public void showDetailsFragment() {
   setFragment(new DetailsFragment(), true);
}
```

File: app/src/main/java/kozak/labs/Fragments/FavoritesFragment.java (hunk 18 above)
Comment (MarkoKuspis) [reply app/src/main/java/kozak/labs/Fragments/FavoritesFragment.java]: 1. Make sure to apply same modification for `ListFragment`
2. Navigate to other fragment inside Presenter#characterSelected()

**Pull request merged**